{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# API Integration using GitHub API to find Bioinformatics Projects on GitHub"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installing Requests\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The requests package allows a Python program to easily request information from a website and examine the response that’s returned. To install requests, issue a command like the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pip install --user requests  --- in the terminal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing an API Response"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we’ll begin to write a program to issue an API call and process the results by identifying the most starred Bioinformatics projects on GitHub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: cs-video-courses\n",
      "Stars: 67951\n",
      "Description: List of Computer Science courses with video lectures.\n",
      "URL: https://github.com/Developer-Y/cs-video-courses\n",
      "--------------------------------------------------\n",
      "Name: dash\n",
      "Stars: 21853\n",
      "Description: Data Apps & Dashboards for Python. No JavaScript Required.\n",
      "URL: https://github.com/plotly/dash\n",
      "--------------------------------------------------\n",
      "Name: bioinformatics\n",
      "Stars: 5737\n",
      "Description: :microscope: Path to a free self-taught education in Bioinformatics!\n",
      "URL: https://github.com/ossu/bioinformatics\n",
      "--------------------------------------------------\n",
      "Name: biopython\n",
      "Stars: 4467\n",
      "Description: Official git repository for Biopython (originally converted from CVS)\n",
      "URL: https://github.com/biopython/biopython\n",
      "--------------------------------------------------\n",
      "Name: deepvariant\n",
      "Stars: 3305\n",
      "Description: DeepVariant is an analysis pipeline that uses a deep neural network to call genetic variants from next-generation DNA sequencing data.\n",
      "URL: https://github.com/google/deepvariant\n",
      "--------------------------------------------------\n",
      "Name: awesome-single-cell\n",
      "Stars: 3281\n",
      "Description: Community-curated list of software packages and data resources for single-cell, including RNA-seq, ATAC-seq, etc.\n",
      "URL: https://github.com/seandavi/awesome-single-cell\n",
      "--------------------------------------------------\n",
      "Name: Awesome-Bioinformatics\n",
      "Stars: 3274\n",
      "Description: A curated list of awesome Bioinformatics libraries and software.\n",
      "URL: https://github.com/danielecook/Awesome-Bioinformatics\n",
      "--------------------------------------------------\n",
      "Name: nextflow\n",
      "Stars: 2833\n",
      "Description: A DSL for data-driven computational pipelines\n",
      "URL: https://github.com/nextflow-io/nextflow\n",
      "--------------------------------------------------\n",
      "Name: ColabFold\n",
      "Stars: 2069\n",
      "Description: Making Protein folding accessible to all!\n",
      "URL: https://github.com/sokrypton/ColabFold\n",
      "--------------------------------------------------\n",
      "Name: fastp\n",
      "Stars: 1993\n",
      "Description: An ultra-fast all-in-one FASTQ preprocessor (QC/adapters/trimming/filtering/splitting/merging...)\n",
      "URL: https://github.com/OpenGene/fastp\n",
      "--------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "# API Endpoint and Query Parameters\n",
    "BASE_URL = \"https://api.github.com/search/repositories\"\n",
    "QUERY = \"bioinformatics\"  # Search term for bioinformatics projects\n",
    "PARAMS = {\n",
    "    \"q\": QUERY,           # Only \"bioinformatics\" keyword\n",
    "    \"sort\": \"stars\",      # Sort by the most stars\n",
    "    \"order\": \"desc\",      # Descending order\n",
    "    \"per_page\": 10        # Number of results per page\n",
    "}\n",
    "\n",
    "def search_repositories():\n",
    "    try:\n",
    "        # Make the GET request\n",
    "        response = requests.get(BASE_URL, params=PARAMS)\n",
    "        response.raise_for_status()  # Raise an error for bad status codes\n",
    "        \n",
    "        # Parse the JSON response\n",
    "        results = response.json()\n",
    "        repositories = results.get(\"items\", [])\n",
    "        \n",
    "        # Display results\n",
    "        for repo in repositories:\n",
    "            print(f\"Name: {repo['name']}\")\n",
    "            print(f\"Stars: {repo['stargazers_count']}\")\n",
    "            print(f\"Description: {repo['description'] or 'No description available'}\")\n",
    "            print(f\"URL: {repo['html_url']}\")\n",
    "            print(\"-\" * 50)\n",
    "    \n",
    "    except requests.exceptions.RequestException as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "\n",
    "# Run the function\n",
    "search_repositories()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Working with the Response Dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the information from the API call stored as a dictionary, we can work with the data stored there"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Status code: 200\n",
      "Total repositories: 34833\n",
      "Repositories returned: 30\n",
      "\n",
      "Number of keys in the first repository: 80\n",
      "\n",
      "Keys in the first repository:\n",
      "allow_forking\n",
      "archive_url\n",
      "archived\n",
      "assignees_url\n",
      "blobs_url\n",
      "branches_url\n",
      "clone_url\n",
      "collaborators_url\n",
      "comments_url\n",
      "commits_url\n",
      "compare_url\n",
      "contents_url\n",
      "contributors_url\n",
      "created_at\n",
      "default_branch\n",
      "deployments_url\n",
      "description\n",
      "disabled\n",
      "downloads_url\n",
      "events_url\n",
      "fork\n",
      "forks\n",
      "forks_count\n",
      "forks_url\n",
      "full_name\n",
      "git_commits_url\n",
      "git_refs_url\n",
      "git_tags_url\n",
      "git_url\n",
      "has_discussions\n",
      "has_downloads\n",
      "has_issues\n",
      "has_pages\n",
      "has_projects\n",
      "has_wiki\n",
      "homepage\n",
      "hooks_url\n",
      "html_url\n",
      "id\n",
      "is_template\n",
      "issue_comment_url\n",
      "issue_events_url\n",
      "issues_url\n",
      "keys_url\n",
      "labels_url\n",
      "language\n",
      "languages_url\n",
      "license\n",
      "merges_url\n",
      "milestones_url\n",
      "mirror_url\n",
      "name\n",
      "node_id\n",
      "notifications_url\n",
      "open_issues\n",
      "open_issues_count\n",
      "owner\n",
      "private\n",
      "pulls_url\n",
      "pushed_at\n",
      "releases_url\n",
      "score\n",
      "size\n",
      "ssh_url\n",
      "stargazers_count\n",
      "stargazers_url\n",
      "statuses_url\n",
      "subscribers_url\n",
      "subscription_url\n",
      "svn_url\n",
      "tags_url\n",
      "teams_url\n",
      "topics\n",
      "trees_url\n",
      "updated_at\n",
      "url\n",
      "visibility\n",
      "watchers\n",
      "watchers_count\n",
      "web_commit_signoff_required\n",
      "\n",
      "Details of the first repository:\n",
      "Name: cs-video-courses\n",
      "Owner: Developer-Y\n",
      "Stars: 67951\n",
      "Description: List of Computer Science courses with video lectures.\n",
      "URL: https://github.com/Developer-Y/cs-video-courses\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "# Make an API call to search for Bioinformatics repositories and store the response.\n",
    "url = 'https://api.github.com/search/repositories?q=bioinformatics&sort=stars'\n",
    "response = requests.get(url)\n",
    "\n",
    "# Check and print the status code to ensure the API call was successful.\n",
    "print(\"Status code:\", response.status_code)\n",
    "\n",
    "# Store the API response in a variable.\n",
    "if response.status_code == 200:  # Ensure the API call succeeded\n",
    "    response_dict = response.json()\n",
    "    \n",
    "    # Print the total number of repositories.\n",
    "    print(\"Total repositories:\", response_dict['total_count'])\n",
    "    \n",
    "    # Explore information about the repositories.\n",
    "    repo_dicts = response_dict['items']\n",
    "    print(\"Repositories returned:\", len(repo_dicts))\n",
    "    \n",
    "    # Examine the first repository.\n",
    "    if repo_dicts:\n",
    "        repo_dict = repo_dicts[0]\n",
    "        print(\"\\nNumber of keys in the first repository:\", len(repo_dict))\n",
    "        \n",
    "        # Print all keys in the first repository, sorted alphabetically.\n",
    "        print(\"\\nKeys in the first repository:\")\n",
    "        for key in sorted(repo_dict.keys()):\n",
    "            print(key)\n",
    "        \n",
    "        # Additional exploration: print specific details of the first repository\n",
    "        print(\"\\nDetails of the first repository:\")\n",
    "        print(f\"Name: {repo_dict['name']}\")\n",
    "        print(f\"Owner: {repo_dict['owner']['login']}\")\n",
    "        print(f\"Stars: {repo_dict['stargazers_count']}\")\n",
    "        print(f\"Description: {repo_dict['description']}\")\n",
    "        print(f\"URL: {repo_dict['html_url']}\")\n",
    "    else:\n",
    "        print(\"No repositories found in the response.\")\n",
    "else:\n",
    "    print(\"Failed to retrieve data from GitHub API.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's pull out some of the keys in repo_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Selected information about the first repository:\n",
      "Name: cs-video-courses\n",
      "Owner: Developer-Y\n",
      "Stars: 67951\n",
      "Repository: https://github.com/Developer-Y/cs-video-courses\n",
      "Created: 2016-10-21T17:02:11Z\n",
      "Updated: 2025-01-27T10:42:35Z\n",
      "Description: List of Computer Science courses with video lectures.\n"
     ]
    }
   ],
   "source": [
    "# Examine and print selected information about the first repository.\n",
    "print(\"\\nSelected information about the first repository:\")\n",
    "try:\n",
    "    print('Name:', repo_dict['name'])\n",
    "    print('Owner:', repo_dict['owner']['login'])\n",
    "    print('Stars:', repo_dict['stargazers_count'])\n",
    "    print('Repository:', repo_dict['html_url'])\n",
    "    print('Created:', repo_dict['created_at'])\n",
    "    print('Updated:', repo_dict['updated_at'])\n",
    "    print('Description:', repo_dict['description'])\n",
    "except KeyError as e:\n",
    "    print(f\"Key error: {e}. Some information may be missing.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summarizing the Top Repositories"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let’s write a loop to print selected information about each of the repositories returned by the API call so we can include them all in the visualization:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Selected information about each repository:\n",
      "\n",
      "Name: cs-video-courses\n",
      "Owner: Developer-Y\n",
      "Stars: 67951\n",
      "Repository: https://github.com/Developer-Y/cs-video-courses\n",
      "Description: List of Computer Science courses with video lectures.\n",
      "\n",
      "Name: dash\n",
      "Owner: plotly\n",
      "Stars: 21853\n",
      "Repository: https://github.com/plotly/dash\n",
      "Description: Data Apps & Dashboards for Python. No JavaScript Required.\n",
      "\n",
      "Name: bioinformatics\n",
      "Owner: ossu\n",
      "Stars: 5737\n",
      "Repository: https://github.com/ossu/bioinformatics\n",
      "Description: :microscope: Path to a free self-taught education in Bioinformatics!\n",
      "\n",
      "Name: biopython\n",
      "Owner: biopython\n",
      "Stars: 4467\n",
      "Repository: https://github.com/biopython/biopython\n",
      "Description: Official git repository for Biopython (originally converted from CVS)\n",
      "\n",
      "Name: deepvariant\n",
      "Owner: google\n",
      "Stars: 3305\n",
      "Repository: https://github.com/google/deepvariant\n",
      "Description: DeepVariant is an analysis pipeline that uses a deep neural network to call genetic variants from next-generation DNA sequencing data.\n",
      "\n",
      "Name: awesome-single-cell\n",
      "Owner: seandavi\n",
      "Stars: 3281\n",
      "Repository: https://github.com/seandavi/awesome-single-cell\n",
      "Description: Community-curated list of software packages and data resources for single-cell, including RNA-seq, ATAC-seq, etc.\n",
      "\n",
      "Name: Awesome-Bioinformatics\n",
      "Owner: danielecook\n",
      "Stars: 3274\n",
      "Repository: https://github.com/danielecook/Awesome-Bioinformatics\n",
      "Description: A curated list of awesome Bioinformatics libraries and software.\n",
      "\n",
      "Name: nextflow\n",
      "Owner: nextflow-io\n",
      "Stars: 2833\n",
      "Repository: https://github.com/nextflow-io/nextflow\n",
      "Description: A DSL for data-driven computational pipelines\n",
      "\n",
      "Name: ColabFold\n",
      "Owner: sokrypton\n",
      "Stars: 2069\n",
      "Repository: https://github.com/sokrypton/ColabFold\n",
      "Description: Making Protein folding accessible to all!\n",
      "\n",
      "Name: fastp\n",
      "Owner: OpenGene\n",
      "Stars: 1993\n",
      "Repository: https://github.com/OpenGene/fastp\n",
      "Description: An ultra-fast all-in-one FASTQ preprocessor (QC/adapters/trimming/filtering/splitting/merging...)\n",
      "\n",
      "Name: biobert\n",
      "Owner: dmis-lab\n",
      "Stars: 1992\n",
      "Repository: https://github.com/dmis-lab/biobert\n",
      "Description: Bioinformatics'2020: BioBERT: a pre-trained biomedical language representation model for biomedical text mining\n",
      "\n",
      "Name: scanpy\n",
      "Owner: scverse\n",
      "Stars: 1983\n",
      "Repository: https://github.com/scverse/scanpy\n",
      "Description: Single-cell analysis in Python. Scales to >1M cells.\n",
      "\n",
      "Name: oneliners\n",
      "Owner: stephenturner\n",
      "Stars: 1892\n",
      "Repository: https://github.com/stephenturner/oneliners\n",
      "Description: Useful bash one-liners for bioinformatics.\n",
      "\n",
      "Name: minimap2\n",
      "Owner: lh3\n",
      "Stars: 1865\n",
      "Repository: https://github.com/lh3/minimap2\n",
      "Description: A versatile pairwise aligner for genomic and spliced nucleotide sequences\n",
      "\n",
      "Name: gatk\n",
      "Owner: broadinstitute\n",
      "Stars: 1755\n",
      "Repository: https://github.com/broadinstitute/gatk\n",
      "Description: Official code repository for GATK versions 4 and up\n",
      "\n",
      "Name: scispacy\n",
      "Owner: allenai\n",
      "Stars: 1739\n",
      "Repository: https://github.com/allenai/scispacy\n",
      "Description: A full spaCy pipeline and models for scientific/biomedical documents.\n",
      "\n",
      "Name: bioconda-recipes\n",
      "Owner: bioconda\n",
      "Stars: 1666\n",
      "Repository: https://github.com/bioconda/bioconda-recipes\n",
      "Description: Conda recipes for the bioconda channel.\n",
      "\n",
      "Name: rust-bio\n",
      "Owner: rust-bio\n",
      "Stars: 1636\n",
      "Repository: https://github.com/rust-bio/rust-bio\n",
      "Description: This library provides implementations of many algorithms and data structures that are useful for bioinformatics. All provided implementations are rigorously tested via continuous integration.\n",
      "\n",
      "Name: bwa\n",
      "Owner: lh3\n",
      "Stars: 1559\n",
      "Repository: https://github.com/lh3/bwa\n",
      "Description: Burrow-Wheeler Aligner for short-read alignment (see minimap2 for long-read alignment)\n",
      "\n",
      "Name: MMseqs2\n",
      "Owner: soedinglab\n",
      "Stars: 1518\n",
      "Repository: https://github.com/soedinglab/MMseqs2\n",
      "Description: MMseqs2: ultra fast and sensitive search and clustering suite\n",
      "\n",
      "Name: galaxy\n",
      "Owner: galaxyproject\n",
      "Stars: 1440\n",
      "Repository: https://github.com/galaxyproject/galaxy\n",
      "Description: Data intensive science for everyone.\n",
      "\n",
      "Name: seqtk\n",
      "Owner: lh3\n",
      "Stars: 1416\n",
      "Repository: https://github.com/lh3/seqtk\n",
      "Description: Toolkit for processing sequences in FASTA/Q formats\n",
      "\n",
      "Name: seqkit\n",
      "Owner: shenwei356\n",
      "Stars: 1352\n",
      "Repository: https://github.com/shenwei356/seqkit\n",
      "Description: A cross-platform and ultrafast toolkit for FASTA/Q file manipulation\n",
      "\n",
      "Name: MultiQC\n",
      "Owner: MultiQC\n",
      "Stars: 1255\n",
      "Repository: https://github.com/MultiQC/MultiQC\n",
      "Description: Aggregate results from bioinformatics analyses across many samples into a single report.\n",
      "\n",
      "Name: getting-started-with-genomics-tools-and-resources\n",
      "Owner: crazyhottommy\n",
      "Stars: 1219\n",
      "Repository: https://github.com/crazyhottommy/getting-started-with-genomics-tools-and-resources\n",
      "Description: Unix, R and python tools for genomics and data science\n",
      "\n",
      "Name: deep_gcns_torch\n",
      "Owner: lightaime\n",
      "Stars: 1144\n",
      "Repository: https://github.com/lightaime/deep_gcns_torch\n",
      "Description: Pytorch Repo for DeepGCNs (ICCV'2019 Oral, TPAMI'2021), DeeperGCN (arXiv'2020) and GNN1000(ICML'2021): https://www.deepgcns.org\n",
      "\n",
      "Name: scipipe\n",
      "Owner: scipipe\n",
      "Stars: 1083\n",
      "Repository: https://github.com/scipipe/scipipe\n",
      "Description: Robust, flexible and resource-efficient pipelines using Go and the commandline\n",
      "\n",
      "Name: graphein\n",
      "Owner: a-r-j\n",
      "Stars: 1057\n",
      "Repository: https://github.com/a-r-j/graphein\n",
      "Description: Protein Graph Library\n",
      "\n",
      "Name: TDC\n",
      "Owner: mims-harvard\n",
      "Stars: 1039\n",
      "Repository: https://github.com/mims-harvard/TDC\n",
      "Description: Therapeutics Commons (TDC-2): Multimodal Foundation for Therapeutic Science\n",
      "\n",
      "Name: csvtk\n",
      "Owner: shenwei356\n",
      "Stars: 1036\n",
      "Repository: https://github.com/shenwei356/csvtk\n",
      "Description: A cross-platform, efficient and practical CSV/TSV toolkit in Golang\n"
     ]
    }
   ],
   "source": [
    "# Summarize selected information about each repository.\n",
    "print(\"\\nSelected information about each repository:\")\n",
    "for repo_dict in repo_dicts:\n",
    "    try:\n",
    "        print('\\nName:', repo_dict.get('name', 'N/A'))\n",
    "        print('Owner:', repo_dict.get('owner', {}).get('login', 'N/A'))\n",
    "        print('Stars:', repo_dict.get('stargazers_count', 'N/A'))\n",
    "        print('Repository:', repo_dict.get('html_url', 'N/A'))\n",
    "        print('Description:', repo_dict.get('description', 'N/A'))\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred while processing repository: {e}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
